### DISCUSSION TITLE
Why are nodes and edges optional?

### DISCUSSION THREAD

**Discussion by typeofweb** [wp-graphql/wp-graphql#1586] [Q&A]
After introspecting the generated schema, we get the following type for Posts:

```graphql
type RootQueryToPostConnection {
  edges: [RootQueryToPostConnectionEdge]
  nodes: [Post]
  pageInfo: WPPageInfo
}
```

Why are all these fields optional instead of required?

```graphql
type RootQueryToPostConnection {
  edges: [RootQueryToPostConnectionEdge!]
  nodes: [Post!]
  pageInfo: WPPageInfo!
}
```

**Comment by jasonbahl** [comment]
@mmiszy Thanks for the question. 

Edges and Nodes are optional because there's no guarantee that there will be data to return. It's possible that no data of that Type has been published yet, or that the data for that connection is private and the requesting user doesn't have proper permission to view the data.

**Reply by typeofweb** [reply]
Makes sense, but wouldn't that just result in an empty array being returned?

**Reply by jasonbahl** [reply]
Yes, which is more ideal than en error being returned. If there was no data to return, but the Schema marked it as required, then the response would include errors, but not having published data of a Type is not an error.

**Reply by typeofweb** [reply]
Maybe I'm missing something, but if you defined the schema as follows:

```graphql
type RootQueryToPostConnection {
  nodes: [Post!]!
}
```

This would still be a valid response:

```json
{
  "nodes": []
}
```

Wouldn't it?

**Reply by jasonbahl** [reply]
@mmiszy, yes, I suppose it would be valid still. 

I've confirmed in WPGraphQL directly. 

![Screen Shot 2020-11-16 at 9 28 05 AM](https://user-images.githubusercontent.com/1260765/99279948-0ab0f100-27ee-11eb-9cb1-a59bb0ac6c47.png)

The general connections Schema has been around for a while in WPGraphQL, so I can't say for certain what all went into the thought process, but I believe I modeled a lot of the WPGraphQL Schema off of Github's design for connections, and they do not have their edges/nodes set to `nonNull`. 

Is there a particular reason you believe this change would be beneficial? Would love to learn more about your experience with requiring edges/nodes vs leaving them nullable.

**Reply by typeofweb** [reply]
I believe it would make DX better, especially when using TypeScript. Currently, to render a list of posts I have to do the following:

```tsx
<ul>
  {data?.posts?.nodes?.map((post) => (
    <li key={post?.id}>{post?.title}</li>
  ))}
</ul>
```

Even though the data is surely there!
